cpp/ql/src/Diagnostics/FailedExtractions.qll

```diff
@@ -0,0 +1,121 @@
+import cpp
+
+/**
+ * The class of errors upon we mark a file as non-successfully extracted.
+ */
+class ReportableError extends Diagnostic {
+  ReportableError() {
+    (
+      this instanceof CompilerDiscretionaryError or
+      this instanceof CompilerError or
+      this instanceof CompilerCatastrophe
+    ) and
+    // If the extractor encounters an error in a compilation, it always emits a
+    // catch-all diagnostic "There was an error during this compilation", to ensure
+    // that the error makes it to the database.
+    // This error doesn't have a file path attached to it, and is thus
+    // useless for us to report. Furthermore, in the common case, we will have a
+    // proper diagnostic for this error we can show.
+    // Instead, we synthesize `TUnknownError` if this is the only error that we can show to the user.
+    not this.getFile().getAbsolutePath() = ""
+  }
+}
+
+private newtype TExtractionError =
+  TReportableError(ReportableError err) or
+  TCompilationFailed(Compilation c, File f) {
+    f = c.getAFileCompiled() and not c.normalTermination()
+  } or
+  // Report generic extractor errors only if we haven't seen any other error-level diagnostic
+  TUnknownError(CompilerError err) { not exists(ReportableError e) }
```

Comment:
I appreciate this comment and the one above, but I still don't get it. Why should the unknown error be hidden when any reportable error, however benign, appears in any file?

Reply:
1. I re-wrote the commend, hopefully providing more clarity
2. I found another DB table we didn't expose in the AST before that allows me to filter a bit better here.